kaitaistruct.lua

```diff
@@ -1,5 +1,14 @@
 local class = require("class")
 local stringstream = require("string_stream")
+local struct = string.unpack and string or require("struct")
```

Comment:
Sorry I know I wrote this but I think what'd be better is:

```suggestion
local unpack = string.unpack or require("struct").unpack
```

Reply:
This would of course mean all other instances in the file must be changed to simply `unpack` instead of `struct.unpack`.

Reply:
I'll work that in, could you just for me, explain why it is better?

Reply:
Sure :)

Here's `string.unpack(foo)` evaluated how Lua would evaluate it:

```lua
local mt = getmetatable(string)
local val = nil
if mt then
    local __index = mt.__index
    if type(__index) == 'table' then
        val = __index.unpack
    elseif type(__index) == 'function' then
        val = __index(string, 'unpack')
    end
end

val(foo)
```

When you do the following:

```lua
local unpack = string.unpack

unpack(foo)
```

then the `unpack(foo)` part just becomes `unpack(foo)` - the indexing bit has been done for you already by caching it at the top.

The point here being that Lua will perform that indexing stuff (in the first code block) in _all places you do `string.unpack(foo)` instead of just once_, even though `string.unpack` is unlikely to change from invocation to invocation.

While most compilers will optimize this lookup for you, Lua does not. LuaJIT will, however - but we can't assume users are using LuaJIT.

So therefore, it's usually better to cache frequently used functions (or functions that are called frequently - e.g. in a tight loop). Most of the `string`/`struct.unpack` and `bit.*` calls are going to be like this.